File: python/src/fancy_sequence.py
# https://leetcode.com/problems/fancy-sequence/
class Fancy:
    def __init__(self):
        self.data = []  # make list as underlying datastructure

        self.operations = []  # make a list to save the todo operation in

    def append(self, val: int) -> None:
        self.data.append(val)  # append to data

    def addAll(self, inc: int) -> None:
        self.operations.append(
            ("+", inc, len(self.data) - 1)
        )  # add "+ inc" to the list of operations todo, apply only on current data

    def multAll(self, m: int) -> None:
        self.operations.append(
            ("*", m, len(self.data) - 1)
        )  # add "* m" to the list of operations todo, apply only on current data

    def getIndex(self, idx: int) -> int:
        if idx >= len(self.data):  # if out of bounds, return -1
            return -1

        value = self.data[idx]  # get value, and perform all operations on it
        for operation, amount, until_index in self.operations:
            if idx > until_index:  # if operation came before data, ignore step
                continue

            if operation == "+":
                value += amount
            else:
                value *= amount

        return value % 1000000007  # return value with modulo 10⁹ + 7

File: rust/src/fancy_sequence.rs
struct Fancy {
    data: Vec<i128>, //i128 for maximum size
}

impl Fancy {
    fn new() -> Self {
        return Fancy { data: vec![] };
    }

    fn append(&mut self, val: i32) {
        self.data.push(val as i128); //add to end of list
    }

    fn add_all(&mut self, inc: i32) {
        for i in 0..self.data.len() { //to every entry, add inc
            self.data[i] += inc as i128;
        }
    }

    fn mult_all(&mut self, m: i32) {
        for i in 0..self.data.len() { // multiply every entry by m
            self.data[i] *= m as i128;
        }
    }

    fn get_index(&self, idx: i32) -> i32 {
        if idx as usize >= self.data.len() { //if out of bounds, return -1
            return -1;
        }

        return (self.data[idx as usize] % 1000000007) as i32; //return value with modulo 10^9 + 7
    }
}
